### DISCUSSION TITLE
Error in if (p.GL > 0.75) { : missing value where TRUE/FALSE needed

### DISCUSSION THREAD

**Discussion by enosazeogie** [ianjonsen/foieGras#52] [General]
Hi Ian,

I'm trying to fit the SSM model to my GPS tracking data (for a seabird species) but I keep getting this error: 
**Error in if (p.GL > 0.75) { : missing value where TRUE/FALSE needed**
I'm not sure what argument I'm not supplying

Here's a subset of my data: 
[Cleaned_GullTrackssubset.csv](https://github.com/ianjonsen/aniMotum/files/13846414/Cleaned_GullTrackssubset.csv)
df <- read.csv("Cleaned_GullTrackssubset.csv", row.names = FALSE)

I have tried various iterations of the arguments:
fitRw <- fit_ssm(df, model = "rw", time.step = NA)
fitCrw<- fit_ssm(df, model = "crw", time.step = NA)
fitMp <- fit_ssm(df, vmax = 5, model = "mp", time.step = NA,
                         control = ssm_control(verbose = 1))

I even tried to force the data into the right format, but still getting that error

d <- format_data(df, date = "date", coord = c("lon","lat"), 
                tz = "UTC")

fit <- fit_ssm(d, model = "crw", time.step = 24)

Could you please help me with figuring out what the problem is?

Thank you,
Esther

**Comment by jamesgrecian** [comment]
Hi Esther,

When using `aniMotum` to process GPS data you need to specify the location class as "G" in the data frame that you are passing to `fit_ssm`. Please check `?fit_ssm` for more info.

I would just reformat your lc column as follows, and then the function should run.

Best,

James

``` r
require(tidyverse)
#> Loading required package: tidyverse
require(aniMotum)
#> Loading required package: aniMotum
#> The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
#> which was just loaded, will retire in October 2023.
#> Please refer to R-spatial evolution reports for details, especially
#> https://r-spatial.org/r/2023/05/15/evolution4.html.
#> It may be desirable to make the sf package available;
#> package maintainers should consider adding sf to Suggests:.
#> The sp package is now running under evolution status 2
#>      (status 2 uses the sf package in place of rgdal)
#> 
#> Attaching package: 'aniMotum'
#> The following object is masked from 'package:purrr':
#> 
#>     map

df <- read_csv("~/Downloads/Cleaned_GullTrackssubset.csv")
#> Rows: 10 Columns: 5
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> dbl  (4): id, lc, lon, lat
#> dttm (1): date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
df <- df |> mutate(lc = "G")
df
#> # A tibble: 10 × 5
#>        id date                lc      lon   lat
#>     <dbl> <dttm>              <chr> <dbl> <dbl>
#>  1 211696 2021-05-25 04:03:28 G     -73.2  40.8
#>  2 211696 2021-05-25 04:18:27 G     -73.2  40.8
#>  3 211696 2021-05-25 04:33:31 G     -73.2  40.8
#>  4 211696 2021-05-25 04:48:29 G     -73.2  40.8
#>  5 211696 2021-05-25 05:03:28 G     -73.2  40.8
#>  6 211696 2021-05-25 05:18:34 G     -73.2  40.8
#>  7 211696 2021-05-25 05:34:05 G     -73.2  40.8
#>  8 211696 2021-05-25 05:48:34 G     -73.2  40.8
#>  9 211696 2021-05-25 06:03:42 G     -73.2  40.8
#> 10 211696 2021-05-25 06:18:31 G     -73.2  40.8

fit <- fit_ssm(df, model = "rw", time.step = 24)
#> fitting rw SSM to 1 tracks...
#>  pars:   -5.74965 -4.71341 1.91815 0 0 0

plot(fit)
#> $`211696`
```

![](https://i.imgur.com/F8d4bBk.png)

<sup>Created on 2024-01-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by enosazeogie** [comment]
Hi James,

I appreciate your help. I've been able to fit the model.

Regards,
Esther

**Comment by enosazeogie** [comment]
Hi,

I downloaded more data, and I can't get the model to fit anymore. 

For the full dataset (dat2), I keep getting this error:
> glimpse(dat2)
Rows: 128,144
Columns: 5
$ id   <chr> "211696", "211696", "211696", "211696", "211696", "211696", "211696", "211696", "211696", "211696", "211…
$ date <dttm> 2021-05-25 04:03:28, 2021-05-25 04:18:27, 2021-05-25 04:33:31, 2021-05-25 04:48:29, 2021-05-25 05:03:28…
$ lc   <chr> "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G",…
$ lon  <dbl> -73.20768, -73.20767, -73.20747, -73.20747, -73.20737, -73.20757, -73.20767, -73.20756, -73.20740, -73.2…
$ lat  <dbl> 40.83309, 40.83319, 40.83356, 40.83337, 40.83317, 40.83350, 40.83343, 40.83396, 40.83349, 40.83374, 40.8…

> fit_crw <- fit_ssm(dat2, model = "crw", time.step = 24)
fitting crw SSM to 9 tracks...
**Error in seq.int(from, by = by, length.out = length.out) : 'length.out' must be a non-negative number**

I tried this
> fit_crw <- fit_ssm(dat2, model = "crw", time.step = NA)
fitting crw SSM to 9 tracks...
 **pars:   1 1 NA NA NA NA NA NA NA NA 0 0 NA 0      Error in nlminb(obj$par, ifelse(control$verbose == 1, myfn, obj$fn), obj$gr,  : 
  NA/NaN gradient evaluation**

**Error in optimHess(par.fixed, obj$fn, obj$gr) : 
  non-finite value supplied by optim
Error in rep$pdHess : $ operator is invalid for atomic vectors**

So I did a subset of the first 50 rows and I got this: 
>df=head(dat2, 50)
>fit <- fit_ssm(df, model = "rw", time.step = 24)
fitting rw SSM to 1 tracks... 
pars: 1.10021 1.46654 2.46752 1.95034 -0.46163 0.41839 
**Warning message: In sqrt(as.numeric(object$diag.cov.random)) : NaNs produced**

And when I try to plot, I get this error (although type=1 plots fine) 
>plot(fit, what = "fitted", type = 2, ask = TRUE)##plots 2D
**Error in FUN(X[[i]], ...) : !anyNA(x) is not TRUE**

I kept increasing the subset. It fit the model for the first 100 rows, but for the first 1000 rows, it gave the same error as for the first 50 rows

Here's the subset of the first 50 rows.
[Cleaned_GullTrackssubset2.csv](https://github.com/ianjonsen/aniMotum/files/13909701/Cleaned_GullTrackssubset2.csv)

I checked and my animotum is the latest version, 1.2. I'm not sure what the problem is, and I'd appreciate your help.

Regards,
Esther

**Comment by jamesgrecian** [comment]
Hi Esther,

When running your data the correlated random walk seems to run fine, but not the random walk. I'm not sure why that would be @ianjonsen? 

Could it be that time step is set to 24 hours when the example data frame spans less than 24 hours? This would result in `aniMotum` trying to generate only one location from the data?

Your example data has a location on average every 15 minutes, and so setting `time.step = .25` seems sensible.

Best,

James


``` r
require(tidyverse)
#> Loading required package: tidyverse
require(aniMotum)
#> Loading required package: aniMotum
#> The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
#> which was just loaded, will retire in October 2023.
#> Please refer to R-spatial evolution reports for details, especially
#> https://r-spatial.org/r/2023/05/15/evolution4.html.
#> It may be desirable to make the sf package available;
#> package maintainers should consider adding sf to Suggests:.
#> The sp package is now running under evolution status 2
#>      (status 2 uses the sf package in place of rgdal)
#> 
#> Attaching package: 'aniMotum'
#> The following object is masked from 'package:purrr':
#> 
#>     map

df <- read_csv("~/Downloads/Cleaned_GullTrackssubset2.csv")
#> Rows: 50 Columns: 5
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (1): lc
#> dbl  (3): id, lon, lat
#> dttm (1): date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

df |> summarise(interval = mean(difftime(date, lag(date)), na.rm = T))
#> # A tibble: 1 × 1
#>   interval     
#>   <drtn>       
#> 1 15.00578 mins

fit <- fit_ssm(df, model = "crw", time.step = .25)
#> fitting crw SSM to 1 tracks...
#>  pars:   0 0 0 0       pars:   0.65492 0.58628 0.41856 0.2284       pars:   0.65702 1.22032 1.18098 0.35766       pars:   -0.31006 2.70153 2.04332 0.82648       pars:   0.22062 2.2577 3.92804 1.07518       pars:   -0.05075 2.48466 2.96426 0.948       pars:   0.03115 2.3236 2.79494 1.90501       pars:   0.20592 2.5047 2.41912 2.78398       pars:   1.13183 2.42618 2.34409 2.63158       pars:   2.04353 2.36026 2.15319 2.95591       pars:   1.84063 2.446 2.31948 3.66558       pars:   1.99794 2.54114 2.39723 4.87055       pars:   2.16676 2.57425 2.39979 5.84701       pars:   2.02625 2.58376 2.46564 7       pars:   2.11117 2.53943 2.3937 7       pars:   2.11117 2.53943 2.3937 7

plot(fit)
#> $`211696`
```

![](https://i.imgur.com/b6ivkXW.png)

<sup>Created on 2024-01-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by ianjonsen** [comment]
@enosazeogie, as @jamesgrecian shows, the issues you have been getting are not `aniMotum` code issues they are likely due to unfortunate `time.step` choices. The best advice I can provide is to try to understand your data by plotting & summarizing (e.g., as @jamesgrecian did) BEFORE you try to fit models to the data! I remind myself to do this almost daily! I am happy to help you overcome issues using `aniMotum` but this thread is better placed in Discussions, so I'm moving it there. 

Best, Ian

**Reply by ianjonsen** [reply]
A common `Warning` message when fit the `rw` SSM to GPS data is the one @enosazeogie received:

```
fit <- fit_ssm(d, model = "rw", time.step = 24)
fitting rw SSM to 1 tracks...
 pars:   1.10021 1.46654 2.46752 1.95034 -0.46163 0.41839      
Warning message:
In sqrt(as.numeric(object$diag.cov.random)) : NaNs produced
```

This even happens when setting the `time.step` to a more reasonable 0.25, as per @jamesgrecian's post:

```
fit <- fit_ssm(d, model = "rw", time.step = 0.25)
fitting rw SSM to 1 tracks...
 pars:   1.71474 1.81085 2.37491 0.26608 -2.57826 0.33292      
Warning message:
In sqrt(as.numeric(object$diag.cov.random)) : NaNs produced
```

In both cases, the warning indicates that standard errors could not be calculated for one or more of the model parameters. You can verify this by looking at a table of parameter estimates. Here's the table for the `rw` model fit with `time.step = 0.25`:

```
fit$ssm[[1]]$par
          Estimate Std. Error
rho_p   0.82978740 0.05847156
sigma_x 5.55521750 0.77636034
sigma_y 6.11566568 0.78701178
rho_o   0.16494149        NaN
tau_x   1.30483947 2.75861013
tau_y   0.07590566        NaN
```

In this case, standard errors could not be calculated for both `rho_o` (Observation error correlation parameter) and `tau_y` (Observation error SD for the `y` direction). This suggests that the `rw` model may not be able to separate these two parameters very well, and that a simpler model might do better. With GPS data, an `rw` model that assumes the `x` and `y` observation errors are independent (ie. the `rho_o` is fixed at 0) is often a better choice. You can specify this using the `map` argument in `fit_ssm`, and you can see that this fits without any warning:

```
fit <- fit_ssm(d, model = "rw", time.step = 0.25, map = list(rho_o = factor(NA)))
fitting rw SSM to 1 tracks...
 pars:   1.70539 1.80836 2.37753 0.32976 -4.58684 

plot(fit, "p")
```
![Rplot](https://github.com/ianjonsen/aniMotum/assets/9984184/c33cddeb-6453-4382-bed5-ba0e440865c7)

**Comment by enosazeogie** [comment]
Thank you for your help, @jamesgrecian and @ianjonsen.

I'm still getting this error, even though I have tried several changes to the code including those you suggested. Considering the first error, I was wondering if there may be an issue with the "control" argument.  

> fit <- fit_ssm(dat2, model = "crw", time.step = NA)
fitting crw SSM to 9 tracks...
 pars:   1 1 NA NA NA NA NA NA NA NA 0 0 NA 0      Error in nlminb(obj$par, ifelse(control$verbose == 1, myfn, obj$fn), obj$gr,  : 
  NA/NaN gradient evaluation

**Error in optimHess(par.fixed, obj$fn, obj$gr) : 
  non-finite value supplied by optim
Error in rep$pdHess : $ operator is invalid for atomic vectors**

> fit <- fit_ssm(dat2, model = "crw", time.step = 0.4) ##0.4 is the average timestep for the full dataset
fitting crw SSM to 9 tracks...
Error in seq.int(from, by = by, length.out = length.out) : 
  'length.out' must be a non-negative number

> fit <- fit_ssm(dat2, model = "crw", time.step = 0.4, map = list(rho_o = factor(NA)))
fitting crw SSM to 9 tracks...
Error in seq.int(from, by = by, length.out = length.out) : 
  'length.out' must be a non-negative number

> fit <- fit_ssm(dat2, model = "rw", time.step = 0.4, map = list(rho_o = factor(NA)))
fitting rw SSM to 9 tracks...
Error in seq.int(from, by = by, length.out = length.out) : 
  'length.out' must be a non-negative number

> fit <- fit_ssm(dat2, model = "rw", time.step = 0.25, map = list(rho_o = factor(NA)))
fitting rw SSM to 9 tracks...
Error in seq.int(from, by = by, length.out = length.out) : 
  'length.out' must be a non-negative number

**Reply by ianjonsen** [reply]
No, it's unlikely an issue with the `control` options as: 1) you are not specifying any so the defaults are being used; and 2) nothing has changed in the underlying code for > 1 year.

```
Error in seq.int(from, by = by, length.out = length.out) :
'length.out' must be a non-negative number
```

This error suggests there's an issue with the dates in your data - either some NA's, duplicate dates (down to the second), or dates out of temporal sequence. I've noticed the 2nd option can happen with some GPS data.

As for the first error, it may be related to the date issue, but if not then I'd either have to see all of your data or a subset that reproduces this specific error.

**Comment by enosazeogie** [comment]
Thank you, Ian. That worked. 

I had previously cleaned out the NAs and duplicates in the latitude, longitude, and time columns, but it turns out there was one sneaky NA in the device ID column which I didn't anticipate. After it took it out, the model ran okay. 

I appreciate your help.

Regards,
Esther
